### PR TITLE
Improve the plumbing and documentation for some of the ask* methods

### PR DESCRIPTION
**Pull request by tyler-romero**
Address a some documentation papercuts for the new ask_* methods. Plus, improve plumbing of arguments to `submit_image_query` and to `wait_for_confident_result`.

### CODE REVIEW

File: src/groundlight/client.py [hunk 1]
```diff
@@ -375,8 +387,10 @@ def ask_confident(
         confidence_threshold: Optional[float] = None,
         wait: Optional[float] = None,
     ) -> ImageQuery:
-        """Evaluates an image with Groundlight waiting until an answer above the confidence threshold
-            of the detector is reached or the wait period has passed.
+        """
+        Evaluates an image with Groundlight waiting until an answer above the confidence threshold
```
Comment (tyler-romero) [inline src/groundlight/client.py:391]: Address some formatting issues
https://code.groundlight.ai/python-sdk/api-reference-docs/#groundlight.client.Groundlight.ask_confident

File: src/groundlight/client.py [hunk 2]
```diff
@@ -469,6 +488,17 @@ def ask_async(
...
+            The longer the patience_time, the more likely Groundlight will arrive at a confident answer.
+            Within patience_time, Groundlight will update ML predictions based on stronger findings,
+            and, additionally, Groundlight will prioritize human review of the image query if necessary.
+            This is a soft server-side timeout. If not set, use the detector's patience_time.
+        :type patience_time: float
+
+        :param confidence_threshold: The confidence threshold to wait for.
+            If not set, use the detector's confidence threshold.
+        :type confidence_threshold: float
+
```
Comment (sunildkumar) [inline src/groundlight/client.py:501]: I absolutely love this description of patience time

File: src/groundlight/client.py [hunk 3]
```diff
@@ -500,26 +530,35 @@ def ask_async(
...
+            # retrieve the result later or on another machine by calling gl.wait_for_confident_result()
+            # with the id of the image_query above. This will block until the result is available.
+            image_query = gl.wait_for_confident_result(image_query.id)
 
             # now the result will be available for your use
             assert image_query.result is not None
 
+            # alternatively, you can check if the result is available (without blocking) by calling
+            # gl.get_image_query() with the id of the image_query above. This will return the image_query,
+            # but the result may still be None.
```
Comment (sunildkumar) [inline src/groundlight/client.py:545]: The result is only None on the servers response to the original ask_async query. At worst the result here is `UNCLEAR` (this is if the best_prediction is None or the placeholder).

File: src/groundlight/client.py [hunk 4]
```diff
@@ -538,10 +578,12 @@ def wait_for_confident_result(
         :return: ImageQuery
         :rtype: ImageQuery
         """
+        if confidence_threshold is None:
+            if isinstance(image_query, str):
+                image_query = self.get_image_query(image_query)
+            confidence_threshold = self.get_detector(image_query.detector_id).confidence_threshold
```
Comment (sunildkumar) [inline src/groundlight/client.py:584]: my only concern here is that this introduces another request from the server. I recall @robotrapta suggesting that we should try to minimize this to minimize overall latency, but maybe I am mistaken. I am guessing that the alternative would be to force the user to provide the detector as an argument to the function which is kinda ugly.

File: src/groundlight/client.py [hunk 5]
```diff
@@ -538,10 +578,12 @@ def wait_for_confident_result(
...
 
-        def confidence_above_thresh(iq):
-            return iq_is_confident(iq, confidence_threshold=confidence_threshold)
-
+        confidence_above_thresh = partial(iq_is_confident, confidence_threshold=confidence_threshold)
```
Comment (sunildkumar) [inline src/groundlight/client.py:586]: thanks for teaching me about `partial` - this is very haskell-y.

File: src/groundlight/client.py [hunk 6]
```diff
@@ -623,6 +662,7 @@ def add_label(self, image_query: Union[ImageQuery, str], label: Union[Label, str
         else:
             image_query_id = str(image_query)
             # Some old imagequery id's started with "chk_"
+            # TODO: handle iqe
```
Comment (sunildkumar) [inline src/groundlight/client.py:665]: I'm not sure what you mean here, so I'm concerned that others might not as well, but maybe I'm just out of the loop.

File: docs/docs/building-applications/5-async-queries.md [hunk 7]
```diff
@@ -67,4 +67,7 @@ result = image_query.result  # This will always be 'None' as you asked asynchron
 
 image_query = gl.get_image_query(id=image_query.id)  # Immediately retrieve the image query from Groundlight
 result = image_query.result  # This will likely be 'UNCLEAR' as Groundlight is still processing your query
+
+image_query = gl.wait_for_confident_result(id=image_query.id)  # Poll for a confident result from Groundlight
+result = image_query.result
```
Comment (sunildkumar) [inline docs/docs/building-applications/5-async-queries.md:72]: I like showing this as the default use case.

File: src/groundlight/client.py (hunk 4 above)
Comment (tyler-romero) [reply src/groundlight/client.py:584]: We actually call get_image_query if image_query is a string as the very first step in `_wait_for_result`, so this just brings the call forwards a bit.

File: src/groundlight/client.py (hunk 3 above)
Comment (sunildkumar) [inline src/groundlight/client.py:545]:
```suggestion
            # but the result may still be UNCLEAR.
```
possible fix...but not the only option

File: src/groundlight/client.py (hunk 5 above)
Comment (tyler-romero) [reply src/groundlight/client.py:586]: Haha its honestly really nice sometimes. I love it because its a bit of a functional holdover

**Review by sunildkumar — APPROVED** [review]
nice! Thanks for your feedback on all of the documentation this afternoon.

File: src/groundlight/client.py (hunk 6 above)
Comment (tyler-romero) [reply src/groundlight/client.py:665]: Edge detectors return image queries with the prefix `iqe_`, so we need to work on a way to handle that sanely

File: src/groundlight/client.py (hunk 4 above)
Comment (sunildkumar) [reply src/groundlight/client.py:584]: ah ok nice

File: src/groundlight/client.py (hunk 6 above)
Comment (brandon-groundlight) [reply src/groundlight/client.py:665]: This what the edge server will return, right? This just opens up the sdk to work with the edge

Comment (sunildkumar) [reply src/groundlight/client.py:665]: got it!

**Review by brandon-groundlight — APPROVED** [review]
Did we also want to attach confidence_threshold to the object sent to the BE, or does that require some BE work first?

File: src/groundlight/client.py (hunk 6 above)
Comment (tyler-romero) [reply src/groundlight/client.py]: Exactly Brandon, just updated the comment to be a bit more clear

**Comment by tyler-romero** [comment]
That will require a bit of BE work Brandon, this is just the intermediate solution

**Pull request merged**